### PR TITLE
C#: Work on cs/unsafe-double-checked-lock

### PR DESCRIPTION
**Pull request by calumgrant**
Tidy up the query and remove some false-positives. Make this query a recommendation as the `volatile` keyword isn't needed on modern .Net runtimes. (Unlike Java, where `volatile` is still required.)

### CODE REVIEW

**Comment by aschackmull** [comment]
> the `volatile` keyword isn't needed on modern .Net runtimes

Really?

File: csharp/ql/test/query-tests/Concurrency/UnsafeLazyInitialization/UnsafeLazyInitialization.cs [hunk 1]
```diff
@@ -39,34 +44,117 @@ static void Main(string[] args)
...
             lock (mutex)
             {
-                if (obj2 == null) { }
+                if (null == obj2)
+                    obj1 = null;
             }
         }
 
-        // BAD: FALSE NEGATIVE - not recognized as double-checked lock
+        // BAD (false-positive): not a double-checked lock
```
Comment (hvitved) [inline csharp/ql/test/query-tests/Concurrency/UnsafeLazyInitialization/UnsafeLazyInitialization.cs]: Isn't this still a false negative?

File: csharp/ql/test/query-tests/Concurrency/UnsafeLazyInitialization/UnsafeLazyInitialization.cs [hunk 2]
```diff
@@ -39,34 +44,117 @@ static void Main(string[] args)
...
+        }
+
+        // BAD: Field x should be volatile
+        if (struct1.x == 2)
+            lock (mutex)
+                if(struct1.x == 2)
+                    struct1.x = 3;
+
+        // GOOD: Tuples are structs so cannot be volatile.
+        if(pair1 == (1,2))
```
Comment (hvitved) [inline csharp/ql/test/query-tests/Concurrency/UnsafeLazyInitialization/UnsafeLazyInitialization.cs]: This is flagged by the query, in contrast to what the comment says.

**Comment by calumgrant** [comment]
It's been really hard to pin this down. The [Wikipedia article](https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Microsoft_.NET_(Visual_Basic,_C#)) claims that the `lock` statement is sufficiently strong and offers enough guarantees. However, I've not found anything more authoritative than that. Yet, many examples and articles still use the `volatile` keyword. So I'm inclined to think that this query is largely redundant - or is only applicable to very old runtimes.

**Comment by aschackmull** [comment]
The wikipedia article claims "In .NET 2.0 and later, 'volatile' is not required." without any reference to back it up, but the `Lazy<T>` implementation still makes use of explicit volatile reads and writes, so I think that's a clear indication that the wikipedia claim is wrong.

**Comment by calumgrant** [comment]
It's possible that `Monitor.Enter` has fewer guarantees than a `lock`
statement, hence the need for explicit volatile reads and writes.

**Comment by calumgrant** [comment]
Code for the volatile field

```
        object makeVolatile()
        {
            if (volatileField is null)
                lock (mutex)
                    if (volatileField is null)
                        volatileField = new object();
            return volatileField;
        }
```

Here is the generated x86 (Release):
```
            if (volatileField is null)
015004CA  in          al,dx  
015004CB  push        edi  
015004CC  push        esi  
015004CD  sub         esp,20h  
015004D0  mov         esi,ecx  
015004D2  lea         edi,[ebp-28h]  
015004D5  mov         ecx,7  
015004DA  xor         eax,eax  
015004DC  rep stos    dword ptr es:[edi]  
015004DE  mov         ecx,esi  
015004E0  mov         dword ptr [ebp-24h],ecx  
015004E3  mov         eax,dword ptr [ebp-24h]  
015004E6  cmp         dword ptr [eax+4],0  
015004EA  jne         01500558  
                lock (mutex)
015004EC  mov         eax,dword ptr [ebp-24h]  
015004EF  mov         edx,dword ptr [eax+0Ch]  
015004F2  mov         dword ptr [ebp-28h],edx  
015004F5  xor         edx,edx  
015004F7  mov         dword ptr [ebp-20h],edx  
015004FA  cmp         byte ptr [ebp-20h],0  
015004FE  je          01500505  
01500500  call        5AE91DFC  
01500505  lea         edx,[ebp-20h]  
01500508  mov         ecx,dword ptr [ebp-28h]  
0150050B  call        733BE300  
                    if (volatileField is null)
01500510  mov         eax,dword ptr [ebp-24h]  
01500513  cmp         dword ptr [eax+4],0  
01500517  jne         01500530  
                        volatileField = new object();
01500519  mov         ecx,5A81EEE0h  
0150051E  call        014A30F4  
01500523  mov         ecx,eax  
01500525  mov         eax,dword ptr [ebp-24h]  
01500528  lea         edx,[eax+4]  
0150052B  call        733BE7E0  
01500530  mov         dword ptr [ebp-18h],0  
01500537  mov         dword ptr [ebp-14h],0FCh  
0150053E  push        1500565h  
01500543  jmp         01500545  
01500545  movzx       eax,byte ptr [ebp-20h]  
01500549  test        eax,eax  
0150054B  je          01500555  
0150054D  mov         ecx,dword ptr [ebp-28h]  
01500550  call        733BE629  
01500555  pop         eax  
01500556  jmp         eax  
            return volatileField;
01500558  mov         eax,dword ptr [ebp-24h]  
            return volatileField;
0150055B  mov         eax,dword ptr [eax+4]  
0150055E  lea         esp,[ebp-8]  
01500561  pop         esi  
01500562  pop         edi  
01500563  pop         ebp  
01500564  ret  
01500565  mov         dword ptr [ebp-14h],0  
0150056C  jmp         01500558  
```

Here is the generated x64 (Release):
```
            if (volatileField is null)
02DF04CA  in          al,dx  
02DF04CB  push        edi  
02DF04CC  push        esi  
02DF04CD  sub         esp,20h  
02DF04D0  mov         esi,ecx  
02DF04D2  lea         edi,[ebp-28h]  
02DF04D5  mov         ecx,7  
02DF04DA  xor         eax,eax  
02DF04DC  rep stos    dword ptr es:[edi]  
02DF04DE  mov         ecx,esi  
02DF04E0  mov         dword ptr [ebp-24h],ecx  
02DF04E3  mov         eax,dword ptr [ebp-24h]  
02DF04E6  cmp         dword ptr [eax+4],0  
02DF04EA  jne         02DF0558  
                lock (mutex)
02DF04EC  mov         eax,dword ptr [ebp-24h]  
02DF04EF  mov         edx,dword ptr [eax+0Ch]  
02DF04F2  mov         dword ptr [ebp-28h],edx  
02DF04F5  xor         edx,edx  
02DF04F7  mov         dword ptr [ebp-20h],edx  
02DF04FA  cmp         byte ptr [ebp-20h],0  
02DF04FE  je          02DF0505  
02DF0500  call        5AE91DFC  
02DF0505  lea         edx,[ebp-20h]  
02DF0508  mov         ecx,dword ptr [ebp-28h]  
02DF050B  call        733BE300  
                    if (volatileField is null)
02DF0510  mov         eax,dword ptr [ebp-24h]  
02DF0513  cmp         dword ptr [eax+4],0  
02DF0517  jne         02DF0530  
                        volatileField = new object();
02DF0519  mov         ecx,5A81EEE0h  
02DF051E  call        00F730F4  
02DF0523  mov         ecx,eax  
02DF0525  mov         eax,dword ptr [ebp-24h]  
02DF0528  lea         edx,[eax+4]  
02DF052B  call        733BE7E0  
02DF0530  mov         dword ptr [ebp-18h],0  
02DF0537  mov         dword ptr [ebp-14h],0FCh  
02DF053E  push        2DF0565h  
02DF0543  jmp         02DF0545  
02DF0545  movzx       eax,byte ptr [ebp-20h]  
02DF0549  test        eax,eax  
02DF054B  je          02DF0555  
02DF054D  mov         ecx,dword ptr [ebp-28h]  
02DF0550  call        733BE629  
02DF0555  pop         eax  
02DF0556  jmp         eax  
            return volatileField;
02DF0558  mov         eax,dword ptr [ebp-24h]  
02DF055B  mov         eax,dword ptr [eax+4]  
02DF055E  lea         esp,[ebp-8]  
02DF0561  pop         esi  
            return volatileField;
02DF0562  pop         edi  
02DF0563  pop         ebp  
02DF0564  ret  
02DF0565  mov         dword ptr [ebp-14h],0  
02DF056C  jmp         02DF0558  
```

Code for the non-volatile field:
```
            if (nonVolatileField is null)
                lock (mutex)
                    if (nonVolatileField is null)
                        nonVolatileField = new object();
            return nonVolatileField;
```

Here is the generated x86 (Release):
```
            if (nonVolatileField is null)
01500582  in          al,dx  
01500583  push        edi  
01500584  push        esi  
01500585  sub         esp,20h  
01500588  mov         esi,ecx  
0150058A  lea         edi,[ebp-28h]  
0150058D  mov         ecx,7  
01500592  xor         eax,eax  
01500594  rep stos    dword ptr es:[edi]  
01500596  mov         ecx,esi  
01500598  mov         dword ptr [ebp-24h],ecx  
0150059B  mov         eax,dword ptr [ebp-24h]  
0150059E  cmp         dword ptr [eax+8],0  
015005A2  jne         01500610  
                lock (mutex)
015005A4  mov         eax,dword ptr [ebp-24h]  
015005A7  mov         edx,dword ptr [eax+0Ch]  
015005AA  mov         dword ptr [ebp-28h],edx  
015005AD  xor         edx,edx  
015005AF  mov         dword ptr [ebp-20h],edx  
015005B2  cmp         byte ptr [ebp-20h],0  
015005B6  je          015005BD  
015005B8  call        5AE91DFC  
015005BD  lea         edx,[ebp-20h]  
015005C0  mov         ecx,dword ptr [ebp-28h]  
015005C3  call        733BE300  
                    if (nonVolatileField is null)
015005C8  mov         eax,dword ptr [ebp-24h]  
015005CB  cmp         dword ptr [eax+8],0  
015005CF  jne         015005E8  
                        nonVolatileField = new object();
015005D1  mov         ecx,5A81EEE0h  
015005D6  call        014A30F4  
015005DB  mov         ecx,eax  
015005DD  mov         eax,dword ptr [ebp-24h]  
015005E0  lea         edx,[eax+8]  
015005E3  call        733BE7E0  
015005E8  mov         dword ptr [ebp-18h],0  
015005EF  mov         dword ptr [ebp-14h],0FCh  
015005F6  push        150061Dh  
015005FB  jmp         015005FD  
015005FD  movzx       eax,byte ptr [ebp-20h]  
01500601  test        eax,eax  
01500603  je          0150060D  
01500605  mov         ecx,dword ptr [ebp-28h]  
01500608  call        733BE629  
0150060D  pop         eax  
0150060E  jmp         eax  
            return nonVolatileField;
01500610  mov         eax,dword ptr [ebp-24h]  
01500613  mov         eax,dword ptr [eax+8]  
01500616  lea         esp,[ebp-8]  
01500619  pop         esi  
0150061A  pop         edi  
0150061B  pop         ebp  
0150061C  ret  
0150061D  mov         dword ptr [ebp-14h],0  
01500624  jmp         01500610  
```

x64 (Release)
```
            if (nonVolatileField is null)
02DF0582  in          al,dx  
02DF0583  push        edi  
02DF0584  push        esi  
02DF0585  sub         esp,20h  
02DF0588  mov         esi,ecx  
02DF058A  lea         edi,[ebp-28h]  
02DF058D  mov         ecx,7  
02DF0592  xor         eax,eax  
02DF0594  rep stos    dword ptr es:[edi]  
02DF0596  mov         ecx,esi  
02DF0598  mov         dword ptr [ebp-24h],ecx  
02DF059B  mov         eax,dword ptr [ebp-24h]  
02DF059E  cmp         dword ptr [eax+8],0  
02DF05A2  jne         02DF0610  
                lock (mutex)
02DF05A4  mov         eax,dword ptr [ebp-24h]  
02DF05A7  mov         edx,dword ptr [eax+0Ch]  
02DF05AA  mov         dword ptr [ebp-28h],edx  
02DF05AD  xor         edx,edx  
02DF05AF  mov         dword ptr [ebp-20h],edx  
02DF05B2  cmp         byte ptr [ebp-20h],0  
02DF05B6  je          02DF05BD  
02DF05B8  call        5AE91DFC  
02DF05BD  lea         edx,[ebp-20h]  
02DF05C0  mov         ecx,dword ptr [ebp-28h]  
02DF05C3  call        733BE300  
                    if (nonVolatileField is null)
02DF05C8  mov         eax,dword ptr [ebp-24h]  
02DF05CB  cmp         dword ptr [eax+8],0  
02DF05CF  jne         02DF05E8  
                        nonVolatileField = new object();
02DF05D1  mov         ecx,5A81EEE0h  
02DF05D6  call        00F730F4  
02DF05DB  mov         ecx,eax  
02DF05DD  mov         eax,dword ptr [ebp-24h]  
02DF05E0  lea         edx,[eax+8]  
02DF05E3  call        733BE7E0  
02DF05E8  mov         dword ptr [ebp-18h],0  
02DF05EF  mov         dword ptr [ebp-14h],0FCh  
02DF05F6  push        2DF061Dh  
02DF05FB  jmp         02DF05FD  
02DF05FD  movzx       eax,byte ptr [ebp-20h]  
02DF0601  test        eax,eax  
02DF0603  je          02DF060D  
02DF0605  mov         ecx,dword ptr [ebp-28h]  
02DF0608  call        733BE629  
02DF060D  pop         eax  
02DF060E  jmp         eax  
            return nonVolatileField;
02DF0610  mov         eax,dword ptr [ebp-24h]  
02DF0613  mov         eax,dword ptr [eax+8]  
02DF0616  lea         esp,[ebp-8]  
02DF0619  pop         esi  
02DF061A  pop         edi  
02DF061B  pop         ebp  
02DF061C  ret  
02DF061D  mov         dword ptr [ebp-14h],0  
02DF0624  jmp         02DF0610  
```
As far as I can tell, there is no difference between the volatile and non-volatile code, on x86_64.

**Comment by aschackmull** [comment]
I'm guessing that the generated CIL will be different (with the loads and stores prefixed by `volatile.`).  Also, once the compiler has chosen an ordering of the instructions semantically consistent with the constraints given by any volatile operations, then since x86 has a quite strong memory model it's plausible that the x86 doesn't need things like one-way memory barriers.  So just because we end up with identical x86 assembly code doesn't allow us to conclude that the volatile had no effect.  Furthermore, architectures like ARM have a much weaker memory model, so compiling to ARM might show you a difference in assembly - as would perhaps another x86 compilation if the code was subtly different, perhaps by doing controlflow splitting on the first `if` to get two separate `return` statements.

File: csharp/ql/src/Concurrency/UnsafeLazyInitialization.qhelp [hunk 3]
```diff
@@ -16,6 +16,7 @@ for example by computing the field twice.
 <li>Make the field volatile using the <code>volatile</code> keyword.</li>
 <li>Use the <code>System.Lazy</code> class, which is guaranteed to be thread-safe. 
 This can often lead to more elegant code.</li>
+<li>Use <code>System.Threading.LazyInitializer</code>.
```
Comment (hvitved) [inline csharp/ql/src/Concurrency/UnsafeLazyInitialization.qhelp]: Missing end `</li>` tag.

File: csharp/ql/src/Concurrency/UnsafeLazyInitialization.qhelp [hunk 4]
```diff
@@ -51,19 +52,26 @@ automatically thread-safe (Recommendation 3):</p>
...
-	</li>
-	<li>
-		MSDN, C# Reference: <a href="https://msdn.microsoft.com/en-us/library/x13ttww7.aspx">volatile</a>.
-	</li>
-	<li>
-		Wikipedia: <a href="https://en.wikipedia.org/wiki/Double-checked_locking">Double-checked locking</a>.
-	</li>
-	
+  <li>
+    MSDN: <a href=https://docs.microsoft.com/en-us/dotnet/api/system.lazy-1">Lazy<T> Class</T></a>.
```
Comment (hvitved) [inline csharp/ql/src/Concurrency/UnsafeLazyInitialization.qhelp]: Missing begin quote.

File: csharp/ql/src/Concurrency/UnsafeLazyInitialization.qhelp [hunk 5]
```diff
@@ -51,19 +52,26 @@ automatically thread-safe (Recommendation 3):</p>
...
-	</li>
-	
+  <li>
+    MSDN: <a href="https://docs.microsoft.com/en-us/dotnet/api/system.lazy-1">Lazy&lt;T&gt; Class</a>.
+  </li>
+  <li>
+    MSDN: <a href="https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized">LazyInitializer.EnsureInitialized Method</a>.
+  </li>
+  <li>
+    MSDN: <a href="https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized">LazyInitializer.EnsureInitialized Method</a>.
```
Comment (jf205) [inline csharp/ql/src/Concurrency/UnsafeLazyInitialization.qhelp]: Same reference appears twice here, so you can delete one of them.

File: change-notes/1.20/analysis-csharp.md [hunk 6]
```diff
@@ -18,7 +18,8 @@
 | Use of default ToString() (cs/call-to-object-tostring) | Fewer false positives | Results have been removed for `char` arrays passed to `StringBuilder.Append()`, which were incorrectly marked as using `ToString`. |
 | Use of default ToString() (cs/call-to-object-tostring) | Fewer results | Results have been removed when the object is an interface or an abstract class. |
 | Unused format argument (cs/format-argument-unused) | Fewer false positives | Results have been removed where the format string is empty. This is often used as a default value and is not an interesting result. |
- 
+| Double-checked lock is not thread-safe (cs/unsafe-double-checked-lock) | Fewer false positives, more true positives | Results have been removed where the underlying field was not updated in the `lock` statement, or where the field is a `struct`. Results have been added where there are other statements inside the `lock` statement. |
```
Comment (jf205) [inline change-notes/1.20/analysis-csharp.md]: All looks fine, but please can you format all of the query `id`s in the table as code, as I think that is the convention (based on looking at earlier versions).

**Comment by jf205** [comment]
All the documentation LGTM, apart from a duplicated reference in the qhelp and a general comment about the change note.
Thanks

**Comment by jf205** [comment]
All the documentation LGTM, apart from a duplicated reference in the qhelp and a general comment about the change note.
Thanks

**Pull request merged**